File: 0. Easy/0136. Single Number/single_number.cpp
// Prompt: https://leetcode.com/problems/single-number/

// This algorithm works because we know that EVERY other number is added twice & only the
// one we want is added once. I'm going to revisit this later & try to do it without using
// extra memory.

// Runtime: 24 ms, faster than 39.44% of C++ online submissions for Single Number.
// Memory Usage: 11.4 MB, less than 15.68% of C++ online submissions for Single Number.

class Solution {
public:
    int singleNumber(vector<int>& nums) {
        std::unordered_set<int> numsFound;
        for (size_t i = 0; i < nums.size(); ++i) {
            if (numsFound.find(nums[i]) == numsFound.end()) {
                numsFound.insert(nums[i]); // add if it's the first time we've seen
            } else {
                numsFound.erase(nums[i]); // remove if it's the second time we've seen it
            }
        }
        return *numsFound.begin(); // the only one left that's not removed is what we need
    }
};

File: 1. Medium/260. Single Number III/single_number.py
# Prompt: https://leetcode.com/problems/single-number-iii/
# Runtime: 44 ms, faster than 69.79% of Python online submissions for Single Number III.
# Memory Usage: 15.1 MB, less than 59.38% of Python online submissions for Single Number III.

class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        numSet = set()
        
        for num in nums:
            if num not in numSet:
                # add every new number to set
                numSet.add(num)
            else:
                # if it's already in the set, remove it
                numSet.remove(num)
        
        # all numbers that were in the list twice will have
        # been added to the set and then removed; the only
        # numbers remaining in the set are the ones that
        # were only in the list once
                
        return list(numSet)
